network-config-analyzer/NetworkLayer.py

```diff
@@ -0,0 +1,221 @@
+        else:
+            ingress_conns = self._allowed_xgress_conns(from_peer, to_peer, True)
+
+        if isinstance(from_peer, IpBlock):
+            egress_conns = PolicyConnections(captured=False, all_allowed_conns=ConnectionSet(True))
+        else:
+            egress_conns = self._allowed_xgress_conns(from_peer, to_peer, False)
+
+        captured_flag = ingress_conns.captured or egress_conns.captured
+        denied_conns = ingress_conns.denied_conns | egress_conns.denied_conns
+        allowed_conns = ingress_conns.all_allowed_conns & egress_conns.all_allowed_conns
+        # captured connections are where at least one of ingress / egress is captured
+        allowed_captured_conns = (ingress_conns.allowed_conns & egress_conns.all_allowed_conns) | \
+                                 (egress_conns.allowed_conns & ingress_conns.all_allowed_conns)
+
+        return allowed_conns, captured_flag, allowed_captured_conns, denied_conns
+
+    def _allowed_xgress_conns(self, from_peer, to_peer, is_ingress):
+        """
+        Implemented by derived classes to get allowed and denied ingress/egress connections between from_peer and to_pee
+        """
+        return NotImplemented
+
+    def collect_policies_conns(self, from_peer, to_peer, is_ingress,
+                               captured_func=lambda policy: True):
+        """
+        Collect allowed/denied/pass connections between two peers, considering all layer's policies that capture the
+        relevant peers.
+        :param Peer.Peer from_peer:  the source peer
+        :param Peer.Peer to_peer: the dest peer
+        :param bool is_ingress: flag to indicate if should return ingress connections or egress connections
+        :param captured_func: callable that returns True if the policy satisfies additional conditions required for
+        considering the target pod as captured and not applying the default connections to it.
+        :return: (allowed_conns, denied_conns, pass_conns, captured_res)
+        :rtype: (ConnectionSet, ConnectionSet, ConnectionSet, bool)
+        """
+        allowed_conns = ConnectionSet()
+        denied_conns = ConnectionSet()
+        pass_conns = ConnectionSet()
+        captured_res = False
+        for policy in self.policies_list:
+            policy_conns = policy.allowed_connections(from_peer, to_peer, is_ingress)
+            if policy_conns.captured:
+                captured_res |= captured_func(policy)
+                policy_conns.denied_conns -= allowed_conns
+                policy_conns.denied_conns -= pass_conns
+                denied_conns |= policy_conns.denied_conns
+                policy_conns.allowed_conns -= denied_conns
+                policy_conns.allowed_conns -= pass_conns
+                allowed_conns |= policy_conns.allowed_conns
+                policy_conns.pass_conns -= denied_conns
+                policy_conns.pass_conns -= allowed_conns
+                pass_conns |= policy_conns.pass_conns
```

Comment:
```suggestion
                policy_conns.denied_conns -= allowed_conns
                policy_conns.denied_conns -= pass_conns
                policy_conns.allowed_conns -= denied_conns
                policy_conns.allowed_conns -= pass_conns
                policy_conns.pass_conns -= denied_conns
                policy_conns.pass_conns -= allowed_conns
                denied_conns |= policy_conns.denied_conns
                allowed_conns |= policy_conns.allowed_conns
                pass_conns |= policy_conns.pass_conns
```
I'm not sure if there is a real example where this makes a difference, but it looks more safe this way

Reply:
done